codegen_llvm/llvm_type: avoid matching on the Rust type

RalfJung opened the pull request:
This `match` is highly suspicious. Looking at `scalar_llvm_type_at` I think it makes no difference. But if it were to make a difference that would be a huge problem, since it doesn't look through `repr(transparent)`!

Cc @eddyb @bjorn3

RalfJung committed:
codegen_llvm/llvm_type: avoid matching on the Rust type

davidtwco reviewed (approved):
This looks okay to me but I want a second pair of eyes on it.

davidtwco commented:
r? @bjorn3

RalfJung commented:
I just found this which seems equally suspicious:

https://github.com/rust-lang/rust/blob/4354192429729d64f205e5021b6e793ea77e65c4/compiler/rustc_codegen_llvm/src/type_of.rs#L308

Moreover, pointee types *are* removed, so the entire hack is probably unnecessary?

RalfJung committed:
also avoid matching on the type in scalar_pair_element_llvm_type

RalfJung committed:
carry out the same changes in the gcc backend

RalfJung commented:
I've removed that other `match` as well now, and since the GCC backend copy-pasted that code I also updated it there.

Cranelift is *extremely* reliant on Rust types in its backend type computation functions, which is concerning... maybe it's just somehow not relevant for ABI there?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6001282823/job/16275115198) [(plain)](https://github.com/rust-lang/rust/commit/99d76a4027fd296118b0e3b55cd4c05a34cf9c08/checks/16275115198/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-nvh53h48/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
<missing>      3 weeks ago          /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      3 weeks ago          /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      3 weeks ago          /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fdf610c9bd0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/faa29d05c2f11d391b1d13208e8e5b291e4cbcfeff006ae78b2a8a6ed45096e6c5ea5d545002779f3e19e30628d9c28d123c4330aeeb14d87fd840fd5cd8cc48
sha256:3c692e31bb7f8c7490708ee97d272c4febfc8b14d40b1fd80f441eb477e35353
---
    Checking smallvec v1.10.0
    Checking gccjit_sys v0.0.1 (https://github.com/antoyo/gccjit.rs#d6e52626)
    Checking gccjit v1.0.0 (https://github.com/antoyo/gccjit.rs#d6e52626)
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error: unused import: `FnAbiOf`
  |
  |
7 | use rustc_middle::ty::layout::{FnAbiOf, LayoutOf, TyAndLayout};
  |
  = note: `-D unused-imports` implied by `-D warnings`

error: unused variable: `immediate`
error: unused variable: `immediate`
   --> src/type_of.rs:268:92
    |
268 |     fn scalar_pair_element_gcc_type<'gcc>(&self, cx: &CodegenCx<'gcc, 'tcx>, index: usize, immediate: bool) -> Type<'gcc> {
    |                                                                                            ^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_immediate`
    = note: `-D unused-variables` implied by `-D warnings`

error: could not compile `rustc_codegen_gcc` (lib) due to 2 previous errors
Build completed unsuccessfully in 0:02:35
```

</details>

bjorn3 commented:
cg_clif should not be dependent on rust types for calling convention handling. It should only use rust types for handling locals.

RalfJung committed:
remove an unused argument

it was already unused before, but due to the recursion the compiler did not realize

RalfJung commented:
Okay, I was probably looking at the wrong code then.

bjorn3 commented:
@bors r+

bors commented:
:pushpin: Commit 9b9cb51a407e46e555afa1c2ec80ced185ad5395 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 9b9cb51a407e46e555afa1c2ec80ced185ad5395 -->
<!-- homu: {"type":"Approved","sha":"9b9cb51a407e46e555afa1c2ec80ced185ad5395","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 9b9cb51a407e46e555afa1c2ec80ced185ad5395 with merge 4e78abb437a0478d1f42115198ee45888e5330fd...
<!-- homu: {"type":"BuildStarted","head_sha":"9b9cb51a407e46e555afa1c2ec80ced185ad5395","merge_sha":"4e78abb437a0478d1f42115198ee45888e5330fd"} -->

eddyb commented on compiler/rustc_codegen_llvm/src/type_of.rs:
```diff
@@ -215,20 +215,16 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyAndLayout<'tcx> {
     /// of that field's type - this is useful for taking the address of
     /// that field and ensuring the struct has the right alignment.
     fn llvm_type<'a>(&self, cx: &CodegenCx<'a, 'tcx>) -> &'a Type {
+        // This must produce the same result for `repr(transparent)` wrappers as for the inner type!
+        // In other words, this should generally not look at the type at all, but only at the
+        // layout.
```
Sorry for seeing this so late, but, if not made clear so far by anyone else:
the special cases you removed are *simultaneously* sound *and* obsolete.

They're sound because they only affect pointer types' pointee types (which doesn't impact ABI on it own), and they're obsolete thanks to LLVM opaque pointers (making the removal welcome!).

You can kinda tell they were working fine because e.g. `Option<&String>` would look like `i64*`, whereas the inner `&String` would actually be `{i8*, i64, i64}*` - worst case, it would pessimize some badly written LLVM optimizations, but all of those concerns should be gone now.

In fact, some of this code can very likely be simplified!

`scalar_pair_element_llvm_type` only exists in the form it does because it needed to special-case wide pointers, but now each component of the scalar pair can be independently converted to LLVM.

It might even make sense to remove caches like `scalar_lltypes` and instead have something much simpler keyed by `Primitive` - the existing `scalar_llvm_type_at` might be good enough, even, and I would argue the cache *only* existed because of the pointer cases, and can now be fully removed.

RalfJung replied:
> the special cases you removed are simultaneously sound and obsolete.

Yes that was my conclusion, too. That still makes this PR a good cleanup, no? :)

But looking at these special cases lead me to ones that aren't sound (see the recent [ABI issues](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AA-abi) I opened), so the cleanup of obsolete-but-sound type matching was still worth it.

I'm sure there's further simplification possible.
